src/vellum/core/context.py

```diff
@@ -0,0 +1,82 @@
+from contextlib import contextmanager
+from functools import wraps
+import inspect
+import threading
+from typing import Any, Callable, Dict, Iterator, TypeVar, Union, cast
+
+from vellum.workflows.events.types import ParentContext
+
+LoggingContext = Dict[str, Any]
+_CONTEXT_KEY = "_execution_context"
+
+local = threading.local()
+
+F = TypeVar('F', bound=Callable[..., Any])
+T = TypeVar('T')
+P = TypeVar('P', bound=ParentContext)
+
+
+def get_execution_context() -> LoggingContext:
+    """Retrieve the current execution context."""
+    return getattr(local, _CONTEXT_KEY, {})
+
+
+def set_execution_context(context: LoggingContext) -> None:
+    """Set the current execution context."""
+    setattr(local, _CONTEXT_KEY, context)
+
+
+@contextmanager
+def execution_context(**kwargs: Any) -> Iterator[None]:
+    """Context manager for handling execution context."""
+    prev_context = get_execution_context()
+    try:
+        set_execution_context({**prev_context, **kwargs})
+        yield
+    finally:
+        set_execution_context(prev_context)
+
+
+def wrapper_execution_parent_context(
```

Comment:
[blocking] I'm having trouble understanding what this annotation enables for us, especially with the type ignores below. I _think_ this is eventually meant to replace/combine with our existing [context](https://github.com/vellum-ai/vellum-python-sdks/blob/main/src/vellum/workflows/state/context.py)?

It looks promising, but I'd love to chat through this pattern synchronously before we commit to it. unfortunately, I don't think I'd be able to until earliest Sunday.

In the meantime, I think we could split out commit 1 from this PR (basically, the minimal changes needed to pass this test: `tests/workflows/basic_parent_context/tests/test_basic_parent_context_workflow_runner.py`). Then we could leave the annotation stuff off for a separate PR until we can chat through it.

wdyt?

Reply:
Heyo! Let's do it. I'm also going to be hard to reach till Sunday. 

I'll split the PR, and harden some of the code here in this PR. 

Thanks for giving a look! Appreciate it!